terminus-ui/popover/src/popover-trigger.directive.ts

```diff
@@ -0,0 +1,302 @@
+    /**
+     * Listen to `keydown` events outside the zone so that change detection is not run every
+     * time a key is pressed. Re-enter the zone only if the `ESC` key is pressed
+     */
+    this.ngZone.runOutsideAngular(() => {
+      (fromEvent<KeyboardEvent>(this.elementRef.nativeElement, 'keydown'))
+        .pipe(
+          filter(event => event.code === KEYS.ESCAPE.code),
+          untilComponentDestroyed(this)
+        ).subscribe(event => this.ngZone.run(() => {
+          this.hide();
+          event.stopPropagation();
+          event.preventDefault();
+        }));
+    });
+    this.document = this.documentService.document;
+
+  }
+
+  /**
+   * When directive initiated, it assigns a reference to popover component and also register on click event.
+   */
+  public ngOnInit(): void {
+    this.popover.referenceObject = this.viewContainerRef.element.nativeElement;
+    this.setContentProperties(this.popover);
+
+    (fromEvent<MouseEvent>(this.elementRef.nativeElement, 'click'))
+      .pipe(
+        untilComponentDestroyed(this)
+      ).subscribe(event => {
+        this.toggle();
+        event.stopPropagation();
+        event.preventDefault();
+      });
+  }
+
+  /**
+   * Called after the default change detector has completed checking all content
+   */
+  public ngAfterContentChecked(): void {
+    if (this.defaultOpened) {
+      this.show();
+    }
+    if (this.popover) {
+      this.popover.myId = this.id;
+    }
+  }
+
+  /**
+   * Called on every change happens
+   * @param changes - SimpleChange
+   */
+  public ngOnChanges(changes: { [propertyName: string]: SimpleChange}): void {
+    if (changes.position) {
+      this.popover.popoverOptions.placement = changes.position.currentValue;
+    }
+  }
+
+  /**
+   * Needed for untilComponentDestroyed
+   */
+  public ngOnDestroy(): void { }
+
+  /**
+   * Set popover options
+   * @param popperRef - TsPopoverComponent
+   */
+  public setContentProperties(popperRef: TsPopoverComponent): void {
+    popperRef.popoverOptions = {
+      ...TsPopoverTriggerDirective.defaultOptions,
+      placement: this.position,
+    };
+    popperRef.onHidden
+      .pipe(untilComponentDestroyed(this))
+      .subscribe(e => this.onPopoverUpdate(e));
+  }
+
+  /**
+   * Toggle between show and hide.
+   */
+  public toggle(): void {
+    this.isOpen ? this.hide() : this.show();
+    this.changeDetectorRef.markForCheck();
+    this.changeDetectorRef.detectChanges();
+  }
+
+  /**
+   * Hide the popover.
+   */
+  public hide(): void {
+    this.isOpen = false;
+    this.popover.hide();
+    this.popoverOnHidden.emit(this.popover.popoverInstance);
+  }
+
+  /**
+   * Show the popover.
+   */
+  public show(): void {
+    this.isOpen = true;
+    this.popover.show(this.popover.popoverOptions);
+    this.popoverOnShown.emit(this.popover.popoverInstance);
+  }
+
+  /**
+   * Emit popoverOnUpdate when it's updated.
+   * @param event
+   */
+  private onPopoverUpdate(event): void {
```

Comment:
We only call this method in one place so we can probably just move the emitter up to where it's used

Reply:
If we do want to leave this method we need to add a type for the event param

Reply:
And add the missing space between the doc comment and the doc param lines